ert_shared/ensemble_evaluator/evaluator.py

```diff
@@ -247,46 +242,69 @@ def stop(self):
         self._loop.call_soon_threadsafe(self._stop)
         self._ws_thread.join()
 
+    def _signal_cancel(self):
+        '''
+        This is just a wrapper around logic for whether to sgnal cancel via
+        a cancellable ensemble or to use internal stop-mechanism directly
+
+        I.e. if the ensemble can be cancelled, it is, otherwise cancel
+        is signalled internally. In both cases the evaluator waits for
+        the  cancel-message to arrive before it shuts down properly.
+        '''
+        if self._ensemble.cancellable:
+            logger.debug("Cancelling current ensemble")
+            self._ensemble.cancel()
+        else:
+            logger.debug("Stopping current ensemble")
+            self._stop()
+
     def run_and_get_successful_realizations(self) -> int:
-        monitor = self.run()
         unsuccessful_connection_attempts = 0
-        while True:
-            try:
-                for _ in monitor.track():
-                    unsuccessful_connection_attempts = 0
-                break
-            except ConnectionClosedError as e:
-                logger.debug(
-                    "Connection closed unexpectedly in "
-                    f"run_and_get_successful_realizations: {e}"
-                )
-            except ConnectionRefusedError as e:
-                unsuccessful_connection_attempts += 1
-                logger.debug(
-                    f"run_and_get_successful_realizations caught {e}."
-                    f"{unsuccessful_connection_attempts} unsuccessful attempts"
-                )
-                if (
-                    unsuccessful_connection_attempts
-                    == _MAX_UNSUCCESSFUL_CONNECTION_ATTEMPTS
-                ):
-                    logger.debug("Max connection attempts reached")
-                    if self._ensemble.cancellable:
-                        logger.debug("Cancelling current ensemble")
-                        self._ensemble.cancel()
+        with self.run() as monitor:
+            while True:
+                try:
+                    for _ in monitor.track():
+                        unsuccessful_connection_attempts = 0
+
+                    logger.debug("Waiting for evaluator shutdown")
+                    break
+                except ConnectionClosedError as e:
+                    logger.debug(
+                        "Connection closed unexpectedly in "
+                        f"run_and_get_successful_realizations: {e}"
+                    )
+                    # The monitor went away...  what the h*** does that mean?
```

Comment:
It means that something prevented the closing handshake. So anything, really. Server crash, connection termination, network issues, etc.

Reply:
👍